### PR TITLE
Convert CppIn/Out/Ref/Var-Ports

### PR DESCRIPTION
**Pull request by seowwj**
Issue Number: 

Objective of pull request: Convert the messaging infrastructure to C++ from Python

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by hexu33 — APPROVED** [review]

File: src/lava/magma/runtime/message_infrastructure/include/ports.h [hunk 1]
```diff
@@ -0,0 +1,208 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+// See: https://spdx.org/licenses/
+
+#ifndef PORTS_H_
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/include/ports.h:5]: Please merge to latest code in dev branch. The macro has been modified to INCLUDE_PORTS_H

File: src/lava/magma/runtime/message_infrastructure/include/ports.h [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    std::vector<PortPtr> ports_;
+};
+
+
+class CppInPort : public AbstractCppIOPort {
+  public:
+    bool Probe();
+    virtual std::vector<> Peek();
+    virtual std::vector<> Recv();
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/include/ports.h:34]: Please add type for vector

File: src/lava/magma/runtime/message_infrastructure/include/ports.h [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+class CppInPortVectorSparse : public CppInPort {
+  public:
+    std::vector<dtype_> Recv();
+    std::vector<dtype_> Peek();
+};
+
+
+class CppInPortScalarDense : public CppInPort {
+  public:
+    std::vector<dtype_> Recv();
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/include/ports.h:60]: Please change dtype following pybind11/numpy.h

File: src/lava/magma/runtime/message_infrastructure/include/ports.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+class CppVarPortScalarSparse : public CppVarPort {
+  public:
+    std::vector<int> Recv();
+    std::vector<int> Peek();
+    void Service(); 
+};
+
+} // namespace message_infrastructure
+
+#endif
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/include/ports.h:208]: Please merge to latest code in dev branch for correct comment style.

File: src/lava/magma/runtime/message_infrastructure/src/ports.cc [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+std::vector<PortPtr> AbstractCppIOPort::GetPorts() {
+  return this->ports_;
+}
+
+bool CppInPort::Probe() {
+  return;
+}
+
+
+}
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/src/ports.cc:19]: }  // namespace message_infrastructure

File: src/lava/magma/runtime/message_infrastructure/include/ports.h (hunk 1 above)
Comment (seowwj) [reply src/lava/magma/runtime/message_infrastructure/include/ports.h]: Updated and ran pylint

**Comment by killight98** [comment]
@seowwj
pls add pythonwrapper interfaces of Ports into `src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/message_infrastructure_py_wrapper.cc`.

File: src/lava/magma/runtime/message_infrastructure/include/ports.h (hunk 1 above)
Comment (seowwj) [reply src/lava/magma/runtime/message_infrastructure/include/ports.h]: @hexu33 CppLint gave an error that the macro should remain as PORTS_H_
https://github.com/lava-nc/lava/runs/7802504468?check_suite_focus=true

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/abstract_port_implementation.cc [hunk 6]
```diff
@@ -22,8 +22,4 @@ std::vector<int> AbstractPortImplementation::GetShape() {
   return this->shape_;
 }
 
-std::vector<PortPtr> AbstractPortImplementation::GetPorts() {
-  return this->ports_;
-}
-
-} // namespace message_infrastructure
\ No newline at end of file
+} // namespace message_infrastrature
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/abstract_port_implementation.cc:25]: Before comments, there should be 2 spaces.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/abstract_port_implementation.h [hunk 7]
```diff
@@ -17,7 +17,7 @@ class AbstractPortImplementation {
   int Start();
   int Join();
   std::vector<int> GetShape();
-  std::vector<PortPtr> GetPorts();
+  virtual std::vector<PortPtr> GetPorts();
 
   DataType dtype_;
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/abstract_port_implementation.h:22]: This data type and data shape could be tranformed following data structure of numpy array.

**Comment by seowwj** [comment]
> @seowwj pls add pythonwrapper interfaces of Ports into `src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/message_infrastructure_py_wrapper.cc`.

@killight98 @hongdami
Question about virtual functions and porting to the PyWrapper: Do we have to define them, or do we leave it out in the wrapper?

**Comment by hexu33** [comment]
> > @seowwj pls add pythonwrapper interfaces of Ports into `src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/message_infrastructure_py_wrapper.cc`.
> 
> @killight98 @hongdami Question about virtual functions and porting to the PyWrapper: Do we have to define them, or do we leave it out in the wrapper?

Classes with virtual functions do not need to be wrapped. Only abstract class has virtual functions and they do not need to be wrapped into Python.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/ports.cc [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+std::vector<int> CppInPortScalarSparse::Peek(){
+}
+
+std::vector <ndarray> CppOutPortVectorDense::Send(ndarray data){
+  for (auto port : this->ports_){
+    port->Send(data);
+  }
+}
+
+}
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/ports.cc:41]: Please add comment, "  // namespace message_infrastructure"

**Comment by killight98** [comment]
@seowwj @shaline-koh 
pls add abstract_port_implementation.cc, ports.cc and transformer.cc into [CMakeLists.txt](https://github.com/lava-nc/lava/blob/messaging_refactor_develop/src/lava/magma/runtime/message_infrastructure/CMakeLists.txt#L7).

**Review by killight98 — APPROVED** [review]

**Pull request merged**